lib/web/command.go

```diff
@@ -20,20 +20,611 @@
+	ExecutionID string `json:"execution_id"`
+}
+
+func (c *CommandRequest) Check() error {
+	if c.Command == "" {
+		return trace.BadParameter("missing command")
+	}
+
+	if c.Login == "" {
+		return trace.BadParameter("missing login")
+	}
+
+	if c.ConversationID == "" {
+		return trace.BadParameter("missing conversation ID")
+	}
+
+	if c.ExecutionID == "" {
+		return trace.BadParameter("missing execution ID")
+	}
+
+	return nil
+}
+
+// executeCommand executes a command on all nodes that match the query.
 func (h *Handler) executeCommand(
 	w http.ResponseWriter,
 	r *http.Request,
 	_ httprouter.Params,
 	sessionCtx *SessionContext,
 	site reversetunnel.RemoteSite,
 ) (any, error) {
-	return nil, trace.NotImplemented("not implemented")
+	q := r.URL.Query()
+	params := q.Get("params")
+	if params == "" {
+		return nil, trace.BadParameter("missing params")
+	}
+	var req *CommandRequest
+	if err := json.Unmarshal([]byte(params), &req); err != nil {
+		return nil, trace.BadParameter("failed to read JSON message: %v", err)
+	}
+
+	if err := req.Check(); err != nil {
+		return nil, trace.BadParameter("invalid payload: %v", err)
+	}
+
+	clt, err := sessionCtx.GetUserClient(r.Context(), site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	identity, err := createIdentityContext(req.Login, sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, err := h.cfg.SessionControl.AcquireSessionContext(r.Context(), identity, h.cfg.ProxyWebAddr.Addr, r.RemoteAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	authAccessPoint, err := site.CachingAccessPoint()
+	if err != nil {
+		h.log.WithError(err).Debug("Unable to get auth access point.")
+		return nil, trace.Wrap(err)
+	}
+
+	netConfig, err := authAccessPoint.GetClusterNetworkingConfig(ctx)
+	if err != nil {
+		h.log.WithError(err).Debug("Unable to fetch cluster networking config.")
+		return nil, trace.Wrap(err)
+	}
+
+	clusterName := site.GetName()
+
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin:     func(r *http.Request) bool { return true },
+	}
+
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		errMsg := "Error upgrading to websocket"
+		h.log.WithError(err).Error(errMsg)
+		http.Error(w, errMsg, http.StatusInternalServerError)
+		return nil, nil
+	}
+
+	defer func() {
+		ws.WriteMessage(websocket.CloseMessage, nil)
+		ws.Close()
+	}()
+
+	keepAliveInterval := netConfig.GetKeepAliveInterval()
+	err = ws.SetReadDeadline(deadlineForInterval(keepAliveInterval))
+	if err != nil {
+		h.log.WithError(err).Error("Error setting websocket readline")
+		return nil, nil
+	}
+
+	hosts, err := findByQuery(ctx, clt, req.Query)
+	if err != nil {
+		log.WithError(err).Warn("failed to find nodes by labels")
+	}
+
+	if len(hosts) == 0 {
+		const errMsg = "no servers found"
+		h.log.Error(errMsg)
+		return nil, trace.Errorf(errMsg)
+	}
+
+	h.log.Debugf("found %d hosts", len(hosts))
```

Comment:
This debug logs seems not very useful without more context. I would either remove it or expand it.
```suggestion
	h.log.Debugf("Found %d hosts to run Assist command %q on.", len(hosts), req.Command)
```